### PR TITLE
Update to 4.0.0

### PR DESCRIPTION
**Pull request by gqmelo**
I've checked out ```cling-patches``` branch [cern repo](http://root.cern.ch/git/llvm.git) and rebased onto ```release_40``` branch [from the official mirror](https://github.com/llvm-mirror/llvm.git).
The result can be seen here: https://github.com/gqmelo/llvm/compare/release_40...gqmelo:cling-patches-4.0
The patches updated here were obtained from this rebased branch.

### CODE REVIEW

File: recipe/cling-patches/0005-Weak-symbols-can-be-functions-too.patch [hunk 1]
```diff
@@ -1,24 +0,0 @@
...
-
-diff --git a/lib/ExecutionEngine/RuntimeDyld/RuntimeDyldELF.cpp b/lib/ExecutionEngine/RuntimeDyld/RuntimeDyldELF.cpp
-index 8065fa3..f2a46b8 100644
---- a/lib/ExecutionEngine/RuntimeDyld/RuntimeDyldELF.cpp
-+++ b/lib/ExecutionEngine/RuntimeDyld/RuntimeDyldELF.cpp
-@@ -1257,6 +1257,7 @@ RuntimeDyldELF::processRelocationRef(
-       Value.Addend = Addend;
-       break;
-     }
-+    case SymbolRef::ST_Function:
```
Comment (gqmelo) [inline recipe/cling-patches/0005-Weak-symbols-can-be-functions-too.patch]: This change is already present on 4.0: https://github.com/llvm-mirror/llvm/blob/release_40/lib/ExecutionEngine/RuntimeDyld/RuntimeDyldELF.cpp#L1015

File: recipe/llvm-lto-static.patch [hunk 2]
```diff
@@ -1,18 +0,0 @@
-LTO is the only shared library built by LLVM.  It breaks llvmlite
```
Comment (gqmelo) [inline recipe/llvm-lto-static.patch]: It was causing conflict, but windows builds seems fine now without this patch.

File: recipe/meta.yaml [hunk 3]
```diff
@@ -16,21 +16,16 @@ source:
   patches:
     # https://llvm.org/bugs/show_bug.cgi?id=15356
     - llvm-elf-relocation.patch
-    # https://llvm.org/bugs/show_bug.cgi?id=25565
-    - llvm-build-race-condition.patch
-    # http://lists.llvm.org/pipermail/llvm-dev/2016-January/094520.html
-    - llvm-lto-static.patch  # [win]
```
Comment (gqmelo) [inline recipe/meta.yaml]: It was causing conflict, but windows builds seems fine now without this patch.

File: recipe/meta.yaml [hunk 4]
```diff
@@ -16,21 +16,16 @@ source:
   patches:
     # https://llvm.org/bugs/show_bug.cgi?id=15356
     - llvm-elf-relocation.patch
-    # https://llvm.org/bugs/show_bug.cgi?id=25565
-    - llvm-build-race-condition.patch
```
Comment (gqmelo) [inline recipe/meta.yaml]: It seems to be solved on 4.0

**Comment by gqmelo** [comment]
Gentle pinging mantainers ;)
@inducer @pitrou @jakirkham @mingwandroid @mattwala @SylvainCorlay

**Comment by inducer** [comment]
Are there corresponding PRs for clang and libcxx?

**Comment by inducer** [comment]
Looks good to me otherwise.

**Comment by gqmelo** [comment]
> Are there corresponding PRs for clang and libcxx?

Not yet, as the packages that depend on llvm pin to 3.9 I thought about building llvm first. Also there are the cling patches applied to clang which are much more numerous and I don't know if it will as easy to port.
But I can try to build them locally before merging this PR.

**Comment by inducer** [comment]
Nah, let's get this moving and see what happens. Thanks for your contribution--and it would be fantastic if you could start on bringing the other packages (especially clang) up to 4.0.

**Comment by gqmelo** [comment]
Thanks!, I'll try to find some time next week to update clang

**Pull request merged**